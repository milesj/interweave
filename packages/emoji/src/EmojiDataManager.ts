import {
	Emoji,
	EMOTICON_OPTIONS,
	fromCodepointToUnicode,
	fromHexcodeToCodepoint,
	generateEmoticonPermutations,
	GroupKey,
	Hexcode,
	Locale,
	MessagesDataset,
	SkinToneKey,
	SubgroupKey,
	TEXT,
} from 'emojibase';
import { CanonicalEmoji } from './types';

const instances = new Map<string, EmojiDataManager>();

export function resetInstances() {
	if (__DEV__) {
		instances.clear();
	}
}

export class EmojiDataManager {
	EMOJIS: Record<string, CanonicalEmoji> = {};

	EMOTICON_TO_HEXCODE: Record<string, Hexcode> = {};

	SHORTCODE_TO_HEXCODE: Record<string, Hexcode> = {};

	UNICODE_TO_HEXCODE: Record<string, Hexcode> = {};

	GROUPS_BY_KEY: Partial<Record<GroupKey, string>> = {};

	SKIN_TONES_BY_KEY: Partial<Record<SkinToneKey, string>> = {};

	SUBGROUPS_BY_KEY: Partial<Record<SubgroupKey, string>> = {};

	data: CanonicalEmoji[] = [];

	flatData: CanonicalEmoji[] = [];

	locale: Locale = 'en';

	version: string = '0.0.0';

	constructor(locale: Locale, version: string) {
		this.locale = locale;
		this.version = version;
	}

	/**
	 * Return or create a singleton instance per locale.
	 */
	static getInstance(locale: Locale, version: string): EmojiDataManager {
		const key = `${locale}:${version}`;

		if (!instances.has(key)) {
			instances.set(key, new EmojiDataManager(locale, version));
		}

		return instances.get(key)!;
	}

	/**
	 * Return dataset as a list.
	 */
	getData(): CanonicalEmoji[] {
		return this.data;
	}

	/**
	 * Return dataset as a flattened list.
	 */
	getFlatData(): CanonicalEmoji[] {
		return this.flatData;
	}

	/**
	 * Package the emoji object with additional data,
	 * while also extracting and partitioning relevant information.
	 */
	packageEmoji(baseEmoji: Emoji): CanonicalEmoji {
		const { emoticon, hexcode, shortcodes = [] } = baseEmoji;
		const emoji: CanonicalEmoji = {
			...baseEmoji,
			canonicalShortcodes: [],
			primaryShortcode: '',
			skins: [],
			unicode: '',
		};

		// Make our lives easier
		if (!emoji.unicode) {
			emoji.unicode = emoji.text && emoji.type === TEXT ? emoji.text : emoji.emoji;
		}

		// Canonicalize the shortcodes for easy reuse
		emoji.canonicalShortcodes = shortcodes.map((code) => `:${code}:`);
		emoji.primaryShortcode = emoji.canonicalShortcodes[0];

		// Support all shortcodes
		emoji.canonicalShortcodes.forEach((shortcode) => {
			this.SHORTCODE_TO_HEXCODE[shortcode] = hexcode;
		});

		// Support all emoticons
		if (emoticon) {
			const emoticons = Array.isArray(emoticon) ? emoticon : [emoticon];

			emoticons.forEach((emo) => {
				generateEmoticonPermutations(emo, EMOTICON_OPTIONS[emo]).forEach((e) => {
					this.EMOTICON_TO_HEXCODE[e] = hexcode;
				});
			});
		}

		// Support all presentations (even no variation selectors)
		this.UNICODE_TO_HEXCODE[fromCodepointToUnicode(fromHexcodeToCodepoint(hexcode))] = hexcode;

		if (emoji.emoji) {
			this.UNICODE_TO_HEXCODE[emoji.emoji] = hexcode;
		}

		if (emoji.text) {
			this.UNICODE_TO_HEXCODE[emoji.text] = hexcode;
		}

		// Map each emoji
		this.EMOJIS[hexcode] = emoji;

		// Apply same logic to all variations
		if (baseEmoji.skins) {
			emoji.skins = baseEmoji.skins.map((skinEmoji) => this.packageEmoji(skinEmoji));
		}

		return emoji;
	}

	/**
	 * Parse and generate emoji datasets.
	 */
	parseEmojiData(data: Emoji[]): CanonicalEmoji[] {
		data.forEach((emoji) => {
			const packagedEmoji = this.packageEmoji(emoji);

			this.data.push(packagedEmoji);
			this.flatData.push(packagedEmoji);

			// Flatten and package skins as well
			if (packagedEmoji.skins) {
				packagedEmoji.skins.forEach((skin) => {
					this.flatData.push(skin);
				});
			}
		});

		return this.data;
	}

	parseMessageData(data: MessagesDataset) {
		if (data.groups) {
			data.groups.forEach((group) => {
				this.GROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.subgroups) {
			data.subgroups.forEach((group) => {
				this.SUBGROUPS_BY_KEY[group.key] = group.message;
			});
		}

		if (data.skinTones) {
			data.skinTones.forEach((skinTone) => {
				this.SKIN_TONES_BY_KEY[skinTone.key] = skinTone.message;
			});
		}
	}
}
