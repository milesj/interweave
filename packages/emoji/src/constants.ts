import { EmojiConfig } from './types';

// We hardcode these values so that newer emojis
// are not displayed for devices that do not support
// them yet. Support cycle is twice a year.
export const MAX_EMOJI_VERSION = 14;
export const LATEST_DATASET_VERSION = '7.0.1';

export const DEFAULT_CONFIG: EmojiConfig = {
	largeSize: '3em',
	renderUnicode: false,
	size: '1em',
};
