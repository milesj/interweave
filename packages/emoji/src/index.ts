/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './constants';
export * from './Emoji';
export * from './EmojiDataManager';
export * from './matchers';
export * from './types';
export * from './useEmojiData';
