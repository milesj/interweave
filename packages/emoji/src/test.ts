import { Locale } from 'emojibase';
import { loadEmojiData, loadMeta, loadShortcodes } from 'emojibase-test-utils';
// eslint-disable-next-line unicorn/import-index
import { CanonicalEmoji, EmojiDataManager } from './index';

export function mockEmojiData(locale: Locale = 'en'): CanonicalEmoji[] {
	const instance = EmojiDataManager.getInstance(locale);

	if (instance.data.length === 0) {
		instance.parseEmojiData(loadEmojiData([loadShortcodes()]));
		instance.parseMessageData(loadMeta());
	}

	return instance.getData();
}
