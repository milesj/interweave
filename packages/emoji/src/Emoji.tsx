/* eslint-disable complexity */

import React, { useMemo } from 'react';
import { EmojiDataManager } from './EmojiDataManager';
import { EmojiProps, Size } from './types';

export function Emoji({
	emoticon,
	enlarge = false,
	hexcode,
	largeSize = '3em',
	path = '{{hexcode}}',
	renderUnicode = false,
	shortcode,
	size = '1em',
	source,
	unicode,
}: EmojiProps) {
	const data = EmojiDataManager.getInstance(source.locale, source.version);
	const styles = useMemo(() => {
		const styles: Record<string, Size | string> = {
			display: 'inline-block',
			verticalAlign: 'middle',
		};

		// Handle large styles
		if (enlarge && largeSize) {
			styles.width = largeSize;
			styles.height = largeSize;

			// Only apply styles if a size is defined
		} else if (size) {
			styles.width = size;
			styles.height = size;
		}

		return styles;
	}, [enlarge, size, largeSize]);

	if (__DEV__ && !emoticon && !shortcode && !unicode && !hexcode) {
		throw new Error(
			'Emoji component requires a `unicode` character, `emoticon`, `hexcode`, or a `shortcode`.',
		);
	}

	// Retrieve applicable unicode character
	let hex = hexcode;

	if (!hex && shortcode) {
		hex = data.SHORTCODE_TO_HEXCODE[shortcode];
	}

	if (!hex && emoticon) {
		hex = data.EMOTICON_TO_HEXCODE[emoticon];
	}

	if (!hex && unicode) {
		hex = data.UNICODE_TO_HEXCODE[unicode];
	}

	// Return the invalid value instead of erroring
	if (!hex || !data.EMOJIS[hex]) {
		return <span>{unicode ?? emoticon ?? shortcode ?? hex}</span>;
	}

	const emoji = data.EMOJIS[hex];

	if (renderUnicode) {
		return <span>{emoji.unicode}</span>;
	}

	// Determine the path
	let src = path || '{{hexcode}}';

	src =
		typeof src === 'function'
			? src(emoji.hexcode, {
					enlarged: enlarge,
					largeSize,
					size: enlarge ? largeSize : size,
					smallSize: size,
			  })
			: src.replace('{{hexcode}}', emoji.hexcode);

	// http://git.emojione.com/demos/latest/sprites-png.html
	// http://git.emojione.com/demos/latest/sprites-svg.html
	// https://css-tricks.com/using-svg/
	return (
		<img
			alt={emoji.unicode}
			aria-label={emoji.label}
			data-emoticon={emoji.emoticon}
			data-hexcode={emoji.hexcode}
			data-shortcodes={emoji.canonicalShortcodes.join(', ')}
			data-unicode={emoji.unicode}
			src={src}
			style={styles}
			title={emoji.label}
		/>
	);
}
