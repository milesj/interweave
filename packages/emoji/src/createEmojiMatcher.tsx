import React from 'react';
import {
	createMatcher,
	InterweaveProps,
	MatcherFactory,
	MatcherFactoryData,
	Node,
	OnMatch,
} from 'interweave';
import { DEFAULT_CONFIG } from './constants';
import { Emoji } from './Emoji';
import { EmojiConfig, EmojiMatch } from './types';

function factory({
	config,
	params,
	props: { emojiSource },
}: MatcherFactoryData<EmojiMatch, InterweaveProps, EmojiConfig>) {
	return <Emoji {...config} {...params} source={emojiSource} />;
}

function onBeforeParse(content: string, { emojiSource }: InterweaveProps): string {
	if (__DEV__ && !emojiSource) {
		throw new Error(
			'Missing emoji source data. Have you loaded with the `useEmojiData` hook and passed the `emojiSource` prop?',
		);
	}

	return content;
}

function onAfterParse(node: Node, { emojiEnlargeThreshold = 1 }: InterweaveProps): Node {
	const content = React.Children.toArray(node);

	if (content.length === 0) {
		return content;
	}

	let valid = false;
	let count = 0;

	// Use a for-loop, as it's much cleaner than some()
	for (const item of content) {
		if (typeof item === 'string') {
			// Allow whitespace but disallow strings
			if (!item.match(/^\s+$/)) {
				valid = false;
				break;
			}
		} else if (React.isValidElement(item)) {
			// Only count towards emojis
			if (item && item.type === Emoji) {
				count += 1;
				valid = true;

				if (count > emojiEnlargeThreshold) {
					valid = false;
					break;
				}

				// Abort early for non-emoji components
			} else {
				valid = false;
				break;
			}
		} else {
			valid = false;
			break;
		}
	}

	if (!valid) {
		return content;
	}

	return content.map((item) => {
		if (!React.isValidElement(item)) {
			return item;
		}

		return React.cloneElement(item, {
			...(item.props as {}),
			enlarge: true,
		});
	});
}

export function createEmojiMatcher(
	pattern: RegExp,
	onMatch: OnMatch<EmojiMatch, InterweaveProps>,
	customFactory: MatcherFactory<EmojiMatch, InterweaveProps, EmojiConfig> = factory,
) {
	return createMatcher<EmojiMatch, InterweaveProps, EmojiConfig>(pattern, customFactory, {
		config: DEFAULT_CONFIG,
		greedy: true,
		onAfterParse,
		onBeforeParse,
		onMatch,
		tagName: 'img',
		void: true,
	});
}
