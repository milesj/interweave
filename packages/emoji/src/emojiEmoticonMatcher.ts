import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import EmojiDataManager from './EmojiDataManager';
import createEmojiMatcher from './createEmojiMatcher';

const EMOTICON_BOUNDARY_REGEX = new RegExp(
  // eslint-disable-next-line no-useless-escape
  `(^|\\\b|\\\s)(${EMOTICON_REGEX.source})(?=\\\s|\\\b|$)`,
);

export default createEmojiMatcher(EMOTICON_BOUNDARY_REGEX, (result, { emojiSource }) => {
  const data = EmojiDataManager.getInstance(emojiSource.locale);
  const emoticon = result.matches[0].trim();

  if (!emoticon || !data.EMOTICON_TO_HEXCODE[emoticon]) {
    return null;
  }

  // Remove padding
  // TODO still needed?
  result.match = String(emoticon);

  return {
    emoticon,
    hexcode: data.EMOTICON_TO_HEXCODE[emoticon],
  };
});
