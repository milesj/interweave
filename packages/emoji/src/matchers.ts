import EMOJI_REGEX from 'emojibase-regex';
import EMOTICON_REGEX from 'emojibase-regex/emoticon';
import SHORTCODE_REGEX from 'emojibase-regex/shortcode';
import { createEmojiMatcher } from './createEmojiMatcher';
import { EmojiDataManager } from './EmojiDataManager';

const EMOTICON_BOUNDARY_REGEX = new RegExp(
	// eslint-disable-next-line no-useless-escape
	`(^|\\\b|\\\s)(${EMOTICON_REGEX.source})(?=\\\s|\\\b|$)`,
);

export const emojiEmoticonMatcher = createEmojiMatcher(
	EMOTICON_BOUNDARY_REGEX,
	(result, { emojiSource }) => {
		const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
		const emoticon = result.matches[0].trim();

		if (!emoticon || !data.EMOTICON_TO_HEXCODE[emoticon]) {
			return null;
		}

		return {
			emoticon,
			hexcode: data.EMOTICON_TO_HEXCODE[emoticon],
			// Remove padding
			match: String(emoticon),
		};
	},
);

export const emojiShortcodeMatcher = createEmojiMatcher(
	SHORTCODE_REGEX,
	(result, { emojiSource }) => {
		const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
		const shortcode = result.matches[0].toLowerCase();

		if (!shortcode || !data.SHORTCODE_TO_HEXCODE[shortcode]) {
			return null;
		}

		return {
			hexcode: data.SHORTCODE_TO_HEXCODE[shortcode],
			shortcode,
		};
	},
);

export const emojiUnicodeMatcher = createEmojiMatcher(EMOJI_REGEX, (result, { emojiSource }) => {
	const data = EmojiDataManager.getInstance(emojiSource.locale, emojiSource.version);
	const unicode = result.matches[0];

	if (!unicode || !data.UNICODE_TO_HEXCODE[unicode]) {
		return null;
	}

	return {
		hexcode: data.UNICODE_TO_HEXCODE[unicode],
		unicode,
	};
});
