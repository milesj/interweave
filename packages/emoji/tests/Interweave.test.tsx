import React from 'react';
import { Interweave, InterweaveProps } from 'interweave';
import { SOURCE_PROP } from 'interweave/test';
import { render } from 'rut-dom';
import { emojiEmoticonMatcher, emojiShortcodeMatcher, emojiUnicodeMatcher } from '../src/matchers';
import { mockEmojiData } from '../src/test';

mockEmojiData('en', '0.0.0');

describe('Interweave (with emoji)', () => {
	it('renders all types', () => {
		const { root } = render<InterweaveProps>(
			<Interweave
				content={
					'This will convert 🐱 \uD83D\uDC36 :man: :3 all 3 emoji types to PNGs and increase the size.'
				}
				emojiEnlargeThreshold={2}
				emojiSource={SOURCE_PROP}
				matchers={[emojiEmoticonMatcher, emojiShortcodeMatcher, emojiUnicodeMatcher]}
				tagName="div"
			/>,
		);

		expect(root.find('img')).toHaveLength(4);
	});

	it('renders emoji shortcode as unicode', () => {
		const { root } = render<InterweaveProps>(
			<Interweave
				content="This has :cat: and :dog: shortcodes."
				emojiSource={SOURCE_PROP}
				matchers={[emojiShortcodeMatcher.extend({ renderUnicode: true })]}
				tagName="div"
			/>,
		);

		expect(root).toMatchSnapshot();
	});

	it('renders emoji unicode (literals) as unicode', () => {
		const { root } = render<InterweaveProps>(
			<Interweave
				content="This has 🐈️ and 🐕️ shortcodes."
				emojiSource={SOURCE_PROP}
				matchers={[emojiUnicodeMatcher.extend({ renderUnicode: true })]}
				tagName="div"
			/>,
		);

		expect(root).toMatchSnapshot();
	});

	it('renders emoji unicode (escapes) as unicode', () => {
		const { root } = render<InterweaveProps>(
			<Interweave
				content={'This has \uD83D\uDC31 and \uD83D\uDC36 shortcodes.'}
				emojiSource={SOURCE_PROP}
				matchers={[emojiUnicodeMatcher.extend({ renderUnicode: true })]}
				tagName="div"
			/>,
		);

		expect(root).toMatchSnapshot();
	});

	it('renders a single emoji enlarged', () => {
		const { root } = render<InterweaveProps>(
			<Interweave
				content=":cat:"
				emojiSource={SOURCE_PROP}
				matchers={[emojiShortcodeMatcher, emojiUnicodeMatcher]}
				tagName="div"
			/>,
		);

		expect(root).toMatchSnapshot();
	});
});
