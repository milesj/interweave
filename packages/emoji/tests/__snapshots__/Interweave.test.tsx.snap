// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interweave (with emoji) renders a single emoji enlarged 1`] = `
<Interweave
  content=":cat:"
  emojiSource={{ compact: false, locale: "en", version: "0.0.0" }}
  matchers={[ <img />, <img /> ]}
  tagName="div"
>
  <Emoji
    enlarge
    hexcode="1F408"
    largeSize="3em"
    shortcode=":cat:"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
  >
    <img
      alt="🐈️"
      aria-label="cat"
      data-hexcode="1F408"
      data-shortcodes=":cat:"
      data-unicode="🐈️"
      src="1F408"
      style={{ display: "inline-block", verticalAlign: "middle", width: "3em", height: "3em" }}
      title="cat"
    />
  </Emoji>
</Interweave>
`;

exports[`Interweave (with emoji) renders emoji shortcode as unicode 1`] = `
<Interweave
  content="This has :cat: and :dog: shortcodes."
  emojiSource={{ compact: false, locale: "en", version: "0.0.0" }}
  matchers={[ <img /> ]}
  tagName="div"
>
  This has 
  <Emoji
    renderUnicode
    hexcode="1F408"
    largeSize="3em"
    shortcode=":cat:"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
  >
    <span>🐈️</span>
  </Emoji>
   and 
  <Emoji
    renderUnicode
    hexcode="1F415"
    largeSize="3em"
    shortcode=":dog:"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
  >
    <span>🐕️</span>
  </Emoji>
   shortcodes.
</Interweave>
`;

exports[`Interweave (with emoji) renders emoji unicode (escapes) as unicode 1`] = `
<Interweave
  content="This has 🐱 and 🐶 shortcodes."
  emojiSource={{ compact: false, locale: "en", version: "0.0.0" }}
  matchers={[ <img /> ]}
  tagName="div"
>
  This has 
  <Emoji
    renderUnicode
    hexcode="1F431"
    largeSize="3em"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
    unicode="🐱"
  >
    <span>🐱</span>
  </Emoji>
   and 
  <Emoji
    renderUnicode
    hexcode="1F436"
    largeSize="3em"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
    unicode="🐶"
  >
    <span>🐶</span>
  </Emoji>
   shortcodes.
</Interweave>
`;

exports[`Interweave (with emoji) renders emoji unicode (literals) as unicode 1`] = `
<Interweave
  content="This has 🐈️ and 🐕️ shortcodes."
  emojiSource={{ compact: false, locale: "en", version: "0.0.0" }}
  matchers={[ <img /> ]}
  tagName="div"
>
  This has 
  <Emoji
    renderUnicode
    hexcode="1F408"
    largeSize="3em"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
    unicode="🐈️"
  >
    <span>🐈️</span>
  </Emoji>
   and 
  <Emoji
    renderUnicode
    hexcode="1F415"
    largeSize="3em"
    size="1em"
    source={{ compact: false, locale: "en", version: "0.0.0" }}
    unicode="🐕️"
  >
    <span>🐕️</span>
  </Emoji>
   shortcodes.
</Interweave>
`;
